File: src/runtime-hw/stub/led.c
#include <runtime-hw/hw.h>
#include <runtime-sys/sys.h>
#include <stdbool.h>
#include <stdint.h>

///////////////////////////////////////////////////////////////////////////////
// LIFECYCLE

/**
 * @brief Initialize a LED unit.
 */
hw_led_t hw_led_init(uint8_t gpio, hw_pwm_t *pwm) {
  // LED not implemented in stub
  (void)gpio; // Suppress unused parameter warning
  (void)pwm;  // Suppress unused parameter warning
  hw_led_t led = {0};
  return led;
}

/**
 * @brief Return the GPIO pin number for the on-board status LED.
 */
uint8_t hw_led_status_gpio() {
  // LED not implemented in stub
  return 0xFF; // Return invalid pin
}

/**
 * @brief Finalize and release an LED.
 */
void hw_led_finalize(hw_led_t *led) {
  // LED not implemented in stub
  sys_assert(led);
  (void)led; // Suppress unused parameter warning
}

///////////////////////////////////////////////////////////////////////////////
// METHODS

/**
 * @brief Check if an LED is valid.
 */
bool hw_led_valid(hw_led_t *led) {
  // LED not implemented in stub
  sys_assert(led);
  (void)led; // Suppress unused parameter warning
  return false;
}

/**
 * @brief Set the LED state.
 */
bool hw_led_set_state(hw_led_t *led, bool on) {
  // LED not implemented in stub
  sys_assert(led);
  (void)led; // Suppress unused parameter warning
  (void)on;  // Suppress unused parameter warning
  return false;
}

/**
 * @brief Set the LED brightness.
 */
bool hw_led_set_brightness(hw_led_t *led, uint8_t brightness) {
  // LED not implemented in stub
  sys_assert(led);
  (void)led;        // Suppress unused parameter warning
  (void)brightness; // Suppress unused parameter warning
  return false;
}

/**
 * @brief Blink the LED continuously.
 */
bool hw_led_blink(hw_led_t *led, uint8_t period_ms, bool repeats) {
  // LED not implemented in stub
  sys_assert(led);
  (void)led;       // Suppress unused parameter warning
  (void)period_ms; // Suppress unused parameter warning
  (void)repeats;   // Suppress unused parameter warning
  return false;
}

/**
 * @brief Fade the LED continuously.
 */
bool hw_led_fade(hw_led_t *led, uint8_t period_ms, bool repeats) {
  // LED not implemented in stub
  sys_assert(led);
  (void)led;       // Suppress unused parameter warning
  (void)period_ms; // Suppress unused parameter warning
  (void)repeats;   // Suppress unused parameter warning
  return false;
}

File: src/Application/LED.m
#include <Application/Application.h>
#include <runtime-hw/hw.h>

@implementation LED

///////////////////////////////////////////////////////////////////////////////
// LIFECYCLE

/**
 * @brief Initialize a pin with a mode.
 */
- (id)init {
  self = [super init];
  if (self == nil) {
    return nil; // Initialization failed
  }

  // Initialize PWM
  uint8_t gpio = hw_led_status_gpio();
  hw_pwm_t pwm;
  if (gpio != 0xFF) {
    pwm = hw_pwm_init(hw_pwm_gpio_unit(gpio), NULL); // Initialize PWM
    if (!hw_pwm_valid(&pwm)) {
#ifdef DEBUG
      sys_printf("Failed to initialize PWM\n");
#endif
      [self release];
      return nil;
    }
  }

  // Initialize LED
  _led = hw_led_init(gpio, &pwm);
  if (!hw_led_valid(&_led)) {
#ifdef DEBUG
    sys_printf("Failed to initialize LED\n");
#endif
    [self release];
    return nil;
  }

  // Set state to "off"
  hw_led_set_state(&_led, false);

  // Return success
  return self;
}

- (void)dealloc {
  hw_led_finalize(&_led);
  [super dealloc];
}

/**
 * @brief Returns the on-board status LED.
 * @return An LED instance if the board exposes a status LED; nil otherwise.
 *
 * Returns an instance of the on-board status LED, which may be simple on/off or
 * able to display different brightness or colors.
 *
 */
+ (LED *)status {
  static LED *status = nil;
  @synchronized(self) {
    if (status == nil) {
      status = [[LED alloc] init];
    }
    return status;
  }
}

///////////////////////////////////////////////////////////////////////////////
// METHODS

/**
 * @brief Set the LED on/off state.
 */
- (BOOL)setState:(BOOL)on {
  sys_assert(hw_led_valid(&_led));
  return hw_led_set_state(&_led, on) ? YES : NO;
}

@end
